---
title: "Exploratory treeinform notebook"
output: html_notebook
---

# Overview

This is a notebook exploring comparisons of Trinity assemblies pre and post treeinform to Drosophila genome CDS and PacBio isoseq assembly.

```{r, setup, include=FALSE}
library(ape)
library(scales)
library(ggplot2)
library(reshape2)
library(dplyr)
library(hutan)
library(treeio)
library(parallel)
library(knitr)
library(runjags)
library(entropy)

cores = detectCores()
if (cores < 1) { cores = 1 }
set.seed(1287234)

PATH="/Users/aguang/Dropbox\ (Brown)/treeinform/ms_treeinform/data/drosophila_analysis/trinity"

source("code/functions.R")
source("code/summary_stats.R")
```

# Data

Trinity isoform to gene data is from:

 * "/Users/aguang/Dropbox (Brown)/treeinform/ms_treeinform/data/drosophila_analysis/trinity"
 
It was generated from our sqlite3 database from agalma using `dump_sqlite.py` with the following command:
`python dump_sqlite.py agalma_trinity.sqlite 0 15 19 23 27 31 35 39`.

```{r, isoform_num, echo=FALSE, warning=FALSE}
thresholds=c(50,5,.5,.005,.0005,.00005,.0000005)
versions=c(15,19,23,27,31,35,39) # treeinform versions
csv_files<-c("trinity.csv",sapply(versions, function(x) paste0("treeinform",x,".csv")))
isoform_genes<-lapply(csv_files, function(x) read.csv(file.path(PATH,x)))
isoform_num<-mclapply(isoform_genes, function(x) cluster_size_distribution(x$gene), mc.cores=cores)

cds<-read.table("/Users/aguang/Dropbox (Brown)/treeinform/drosophila_analysis/dmel_unique_protein_isoforms_fb_2018_04.tsv",sep="\t",skip=4,col.names=c("FBgn","FB_gene_symbol","representative_protein","identical_protein(s)"),header=FALSE,fill=TRUE)
cds_isoform_num<-cluster_size_distribution(cds$FBgn)
```

```{r, trees_before}
```

```{r, gapped_histogram}
plot(cars)
```

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Cmd+Option+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Cmd+Shift+K* to preview the HTML file).
