---
title: Revising transcriptome assemblies with phylogenetic information in Agalma1.0
  - Figure 1
output:
  pdf_document: default
  html_notebook: default
  html_document: default
  theme: classic
---

```{r, setup, include=FALSE}
library(ape)
library(treeio)
library(ggtree)

library(dplyr)
library(parallel)
library(phangorn)
library(Biostrings)
library(BiocGenerics)

cores = detectCores()
if (cores < 1) { cores = 1 }
set.seed(1287234)
```

Running this R notebook will regenerate Figure 1 from the manuscript. Data consists of:

- **data/revisions/phyldog/before/7.ReconciledTree**: Gene tree generated by Agalma1.1.0 and phyldog before running treeinform.
- **data/revisions/7.fa**: Multiple sequence alignment 7 from Agalma1.1.0 which corresponds to gene tree 7. This gene tree was selected through a manual search through the set of pre-treeinform gene trees for a calde of genes from the same species with improbably short terminal branches.

```{r, fig1a, echo=FALSE, include=FALSE}
fig1a_tree = read.nhx("data/revisions/phyldog/before/7.ReconciledTree")@phylo
fig1a_labels = unlist(strsplit(fig1a_tree$tip.label, "@"))
fig1a_colors = rep(1, length(fig1a_tree$tip.label)+fig1a_tree$Nnode)
#32566, 32567, 32568, 32569 are the model IDs we're looking for
# here they are 30,31,32,33
fig1a_colors[30:33] = 2
hydra_ancestor = getMRCA(fig1a_tree, 30:33)
fig1a_colors[hydra_ancestor] = 2

fig1a_tree$tip.label = gsub("@", ".", fig1a_tree$tip.label)

phyDat = read.phyDat("data/revisions/7.fa", format="fasta", type="AA")
fit = pml(fig1a_tree, phyDat)
#anc.ml = ancestral.pml(fit, "ml")
anc.ml = ancestral.pml(fit, "ml", return="phyDat")
# root is 34
test = as.character(anc.ml)
anc.seq = test[34,]
diff = sweep(test, 2, anc.seq, "==")

test_sum = apply(diff*1, 2, sum)
test_order = order(test_sum, decreasing=TRUE)
ordered_diff = diff[,test_order]

test_diff = apply(ordered_diff, c(1,2), function(x) if(x) { return("A") } else { return("C") })
test_DNAbin = as.DNAbin(test_diff)
write.dna(test_DNAbin, "data/revisions/modified.fasta", format="fasta", nbcol=-6, colsep='')

label = sapply(strsplit(sapply(strsplit(fig1a_tree$tip.label, "\\."), function(x) x[1]), "_"), function(x) x[1])

new_tree = chronos(fig1a_tree)
class(new_tree) = "phylo"
```

```{r, plot, echo=FALSE, warning=FALSE, fig.cap='Gene tree 7 in before-treeinform set with associated alignment. Note that for Hydra_magnipapillata@32566, 32567, 32568, 32569, that 32566 and 32567 share the exact same sequences while the pairs (32567, 32568) and (32567, 32569) each differ only by a small gap. (this was additionally checked with seqinr::dist.alignment and ape::dist.gene). This suggests that Trinity/2.5.2 does not in fact address the transcript misassignment error.'}
#tiff(filename="Figures/figure1.tiff",width=7,height=5,units="in",pointsize=8,res=350)
#postscript(file="Figures/figure1.eps",width=7,height=5,fonts=c("sans"))
#cairo_ps(file="Figures/figure1.eps",width=7,height=5,pointsize=8)
msaplot(ggtree(new_tree), "data/revisions/modified.fasta", color=c("white","black","white","white"), offset=0.4, width=0.8) + geom_tiplab(size=5, label=label) + geom_hilight(node=hydra_ancestor, extend=0.25, fill='grey')
#dev.off()
suppressMessages(msaplot(ggtree(tree)+geom_tiplab(size=2,align=TRUE,linesize=.5),fasta="7.fa",offset=0.7))
```

```{r, for_presentation, eval=FALSE, include=FALSE}
# CODE IN HERE FOR SLIDE PRESENTATIONS ONLY

# problem in ggtree, for presentation
#ggtree(new_tree, color="white") + geom_tiplab(size=5, aes(color=fig1a_colors), label=label) + theme_tree(bgcolor="#222222", fgcolor="white") + xlim(NA, 1.1) + scale_color_continuous(low="white",high="#34A5DA")

# with MSA
# msaplot(ggtree(new_tree, color="white"), "data/modified.fasta", color=c("#222222","gray","#222222","#222222"), offset=0.3, width=0.8) + geom_tiplab(size=5, aes(color=fig1a_colors), label=label) + theme_tree(bgcolor="#222222", fgcolor="white") + scale_color_continuous(low="white",high="#34A5DA")
```
