---
title: Revising transcriptome assemblies with phylogenetic information in Agalma1.0
  - Figure 1
output:
  pdf_document: default
  html_notebook: default
  html_document: default
  theme: classic
---

```{r, setup, include=FALSE}
library(ape)
library(dplyr)
library(parallel)
library(ggtree)
library(phangorn)
library(Biostrings)
library(BiocGenerics)

cores = detectCores()
if (cores < 1) { cores = 1 }
set.seed(1287234)
```

Running this R notebook will regenerate Figure 1 from the manuscript. Data consists of:

- **data/7taxa_original_headers.trees**: Gene trees generated by Agalma1.0 before running treeinform.
- **data/7taxa_treeinform_histo.out**: Output generated by treeinform in the form of a text file where each line consists of a clade of transcripts assigned to the same gene, structured as \$subtree_height \$Agalma IDs of transcripts
- **data/91_alignment.fa**: Multiple sequence alignment 91 from Agalma1.0, which corresponds to gene tree 91 from Agalma1.0, or gene tree 91 in **7taxa_original_headers.trees**. This gene tree was selected through a manual search through **7taxa_treeinform_histo.out** for a clade of genes from the same species with improbably short terminal branches.

```{r, fig1a, echo=FALSE, include=FALSE}
original_trees = read.tree("data/7taxa_original_headers.trees")
original_labels = mclapply(original_trees, function(x) unlist(strsplit(x$tip.label, "@")))
fig1a_tree_ind = which(sapply(original_labels, function(x) any(17744==x) & any(17745==x) & any(17743==x)))
fig1a_tree = original_trees[[fig1a_tree_ind]]
fig1a_colors = rep(1, length(fig1a_tree$tip.label)+fig1a_tree$Nnode)
fig1a_colors[9:11] = 2
hydra_ancestor = getMRCA(fig1a_tree, 9:11)
fig1a_colors[hydra_ancestor] = 2

fig1a_tree$tip.label = gsub("@", ".", fig1a_tree$tip.label)

phyDat = read.phyDat("data/91_alignment.fa", format="fasta", type="AA")
fit = pml(fig1a_tree, phyDat)
#anc.ml = ancestral.pml(fit, "ml")
anc.ml = ancestral.pml(fit, "ml", return="phyDat")
# root is 25
test = as.character(anc.ml)
anc.seq = test[25,]
diff = sweep(test, 2, anc.seq, "==")

test_sum = apply(diff*1, 2, sum)
test_order = order(test_sum, decreasing=TRUE)
ordered_diff = diff[,test_order]

test_diff = apply(ordered_diff, c(1,2), function(x) if(x) { return("A") } else { return("C") })
test_DNAbin = as.DNAbin(test_diff)
write.dna(test_DNAbin, "data/modified.fasta", format="fasta", nbcol=-6, colsep='')

label = sapply(strsplit(sapply(strsplit(fig1a_tree$tip.label, "\\."), function(x) x[1]), "_"), function(x) x[1])

new_tree = chronos(fig1a_tree)
class(new_tree) = "phylo"

# order character matrix by most conserved to least conserved
```

```{r, plot, echo=FALSE, warning=FALSE}
msaplot(ggtree(new_tree), "data/modified.fasta", color=c("white","black","white","white"), offset=0.3, width=0.8) + geom_tiplab(size=5, aes(color=fig1a_colors), label=label)


```