---
title: "Revising transcriptome assemblies with phylogenetic information in Agalma1.0 - Supplementary Materials"
output:
  pdf_document:
    toc: true
  html_notebook: default
  html_document: default
bibliography: supplementary.bib
---

```{r, setup, include=FALSE}
library(ape)
library(ggplot2)
library(dplyr)
library(hutan)
library(treeio)
library(parallel)
library(knitr)
library(runjags)

cores = detectCores()
if (cores < 1) { cores = 1 }
set.seed(1287234)

source("code/functions.R")
```

```{r, parse_trees, include=FALSE, warnings=FALSE}
# Parse trees before everything else
# make speciation calibration times (taken from chronos)
age <- c(0.1718036, 0.1430510, 0.2490215, 0.3227746, 0.5157844, 1.0000000)
clade <- c("Calycophora", "Agalmatidae", "Chodorophora", "Siphonophora", "Hydrozoa", "Cnidaria")
calibration_times = data.frame(age, clade, stringsAsFactors = FALSE)

#5299 gene trees + 5301 (missing 5300)
s <- 1:260
#before <- dt_phyldog(s, "data/before/", calibration_times)
#trees_before <- before[[1]]
#calibrated_before <- before[[2]]
#dt_before <- before[[3]]
#save(dt_before, file="data/dt_before.Rdata")
#save(trees_before, file="data/trees_before.Rdata")
#save(calibrated_before, file="data/calibrated_before.Rdata")
load("data/dt_before.Rdata")
load("data/trees_before.Rdata")
load("data/calibrated_before.Rdata")

#k <- 7000:7300
#threshold_5 <- dt_phyldog(k, "data/threshold-5/", calibration_times)
#save(threshold_5, file="data/dt_threshold-5.Rdata")[[3]]
load("data/dt_threshold-5.Rdata")

#l <- 27804:28104
#threshold_2 <- dt_phyldog(l, "data/threshold-2/", calibration_times)
#save(threshold_2, file="data/dt_threshold-2.Rdata")[[3]]
load("data/dt_threshold-2.Rdata")

#threshold_8 <- dt_phyldog(l, "data/threshold-8/", calibration_times)[[3]]
#save(threshold_8, file="data/dt_threshold-8.Rdata")
load("data/dt_threshold-8.Rdata")

#threshold_mix <- dt_phyldog(l, "data/threshold-mix/", calibration_times)[[3]]
#save(threshold_mix, file="data/dt_threshold-mix.Rdata")
load("data/dt_threshold-mix.Rdata")
```

# Threshold selection and validation

```{r, subtree_histo, echo=FALSE, warning=FALSE, fig.cap="We plotted a histogram of subtree lengths for internal nodes in each gene tree from Agalma1.0 containing tip descendants from the same species. We filtered subtree lengths greater than 1 out for clarity. Over 300 clades of genes from the same species have a subtree length of close to 0. It is unlikely that all of these clades are gene duplication events. From this histogram, we visually selected 0.05 based on the fact that bin counts were very high under that threshold, with a noticeable gap after that threshold."}
lengths = mclapply(trees_before, function(x) {
  phylo = x@phylo
  nnode = phylo$Nnode
  data = x@data
  bl = c(phylo$edge.length, 0)
  data[,'bl'] = bl[order(c(phylo$edge[,2], length(phylo$tip.label)+1))]
  des = lapply(length(phylo$tip.label)+1:nnode, function(z) descendants(phylo, z))
  td = lapply(length(phylo$tip.label)+1:nnode, function(z) tip_descendants(phylo, z))
  duplicates = sapply(td, function(z) {
      sapply(strsplit(names(z), '@'), '[')[1,] %>% duplicated %>% any
    })
  only_duplicates = des[which(duplicates)]
  len = lapply(only_duplicates, function(z) {
      filter(data, node %in% z) %>% select(bl) %>% sum
    })
  return(len)
  }, mc.cores=cores)

lengths_lim = unlist(lengths)[which(unlist(lengths)<1)]

#heights = read.table("data/7taxa_treeinform_histo.out", sep=" ", fill=TRUE, header=FALSE, colClasses = c("numeric", rep("integer", 15))) # don't actually know what max clade size is but assuming it's 15
#colnames(heights) <- c("height", sapply(1:15, function(x) paste0("ID",x)))
ggplot(data=data.frame(x=lengths_lim), aes(x)) + geom_histogram(binwidth=0.02) + theme_classic() + xlab("Subtree length") + ylab("Count") + scale_x_sqrt()
```

To more closely examine threshold selection, we ran phyldog [@boussau2013genome]

further justify a threshold, 

 Second, we ran phyldog  on the multiple sequence alignments from Agalma1.0 and a user-inputted species tree in order to infer gene trees with internal nodes annotated as duplication or speciation events. We then fitted chronograms onto these gene trees with our user-inputted species tree. This allowed us to fit a mixture model of...

In order to select a threshold, we additionally computed a mixture model. Since transcripts from the same gene assigned to different genes bias duplication times towards 0,

and can be modeled as a gamma distribution, we can view the empirical distribution of duplication times before treeinform as a 2-component mixture of the gamma distribution and the theoretical provided by Gernhard:

$$ P(x_{i,k}) = \pi_1 \Gamma(x_{i,k}|\alpha, \beta) + \pi_2 f(x_{i,k} | t_{or,k}=t, \lambda, \mu)$$
where $\Gamma(x_{i,k} | \alpha, \beta)$ is the pdf for the gamma distribution with shape $\alpha$ and rate $\beta$, and $f(x_{i,k} | t_{or,k} = t, \lambda, \mu)$ is the pdf for the duplication times under birth rate $\lambda$ and death rate $\mu$ from above. $\pi_1$ and $\pi_2$ denote the mixing proportions, thus $\pi_1 + \pi_2 = 1$.

```{r, runjags, include=FALSE}
#load("data/dt_before.Rdata")

# Data
#Y = dt_before
#N = length(dt_before)

# Initial values to get the chains started:
#alpha <- 1
#beta <- 1
#lambda <- 0.637025240076
#mu <- 0.614249271090
#t <- 1
#Constant <- 10

#Ones <- rep(1, N)

#results <- run.jags(model="code/runjags.txt", n.chains=3, thin=1)
#save(results, file="data/runjags_results.Rdata")
load("data/runjags_results.Rdata")
sum_results = results$summaries

alpha = sum_results[1,4]
beta = sum_results[2,4]
mu = sum_results[3,4]
lambda = sum_results[4,4]
p1 = sum_results[5,4]
p2 = sum_results[6,4]

mix1 = function(x, p1=p1, alpha=alpha, beta=beta) { p1*dgamma(x,shape=alpha,scale=beta)}
mix2 = function(x, p2=p2, lambda=lambda,mu=mu) { p2*dup_pdf(x,1,lambda,mu)}
```

We use Bayesian Gibbs Sampling with the Bernoulli Ones trick to infer the parameters $\alpha$, $\beta$, $\lambda$, and $\mu$ as well as the mixing proportions $p_1$ and $p_2$. The package we use is Just Another Gibbs Sampler (JAGS).

From JAGS we get the parameter estimates as:
```{r, runjags_table, echo=FALSE}
sum_format = sum_results[,c("Lower95", "Median", "Upper95", "Mean", "MCerr")]
rownames(sum_format) <- c("$\\alpha$", "$\\beta$", "$\\mu$", "$\\lambda$", "$\\pi_1$", "$\\pi_2$")
kable(sum_format)
```

Plotted onto a histogram of the inferred duplication times, we see that most of the duplication times attributed to splice variants are < 0.05 with the actual intersection point between the two distribution curves being 0.009799746. When we back-calibrated the intersection point, we get 0.0208. This intersection point suggests that a threshold choice of around 0.05 is appropriate. (NOTE: duplication times is not the same as subtree height since it is calibrated vs uncalibrated, so the times are not directly comparable with threshold choice)
```{r, runjags_viz, echo=FALSE, warnings=FALSE}
ggplot(data.frame(x=dt_before)) + geom_histogram(aes(x=x,y=..density..), fill='white', color='black', binwidth=0.01) + stat_function(geom="line",fun=mix1,args=list(p1,alpha,beta),xlim=c(0.005,1), color="#FF000080")+stat_function(geom="line",fun=mix2,args=list(p2,lambda,mu),color="blue") + ggtitle("Density Curves of Mixture Model Plotted on Histogram of\n Inferred Duplication Times Before Treeinform") + xlab("Duplication Times") + ylab("Density") + theme_classic()
```

```{r, back_calibrate}
intersection = uniroot(function(x) mix1(x,p1,alpha,beta)-mix2(x,p2,lambda,mu), c(0,0.25))$root

# find duplication events with t<=intersection in calibrated trees
calibrated_thresh = mclapply(calibrated_before, function(x)
  {
  if(is.na(x)) { return(NA) }
  else {
    l = length(x@phylo$tip.label)
    filter(x@data, height <= intersection, node > l) %>% select(node) %>% unlist }
  }, mc.cores = cores)

# compute subtree length for equivalent event in uncalibrated trees
subtree_lengths = mcmapply(function(x, y) {
  if(length(x)==0 || is.na(x)) { return(0) }
  else {
    phylo = y@phylo
    data = y@data
    bl = c(phylo$edge.length, 0)
    data[,'bl'] = bl[order(c(phylo$edge[,2], length(phylo$tip.label)+1))]
    des = lapply(x, function(z) descendants(phylo, z))
    len = lapply(des, function(z) {
      filter(data, node %in% z) %>% select(bl) %>% sum
    })
    return(len)
  }
}, calibrated_thresh, trees_before, mc.cores=cores)

# select max out of subtree lengths
thresh = unlist(subtree_lengths) %>% max
```

# Figure 3: Threshold selection analysis (Validation)

## Figure 3a: Percentage of reassigned tips 

The percentage of reassigned tips is plotted below. The original assembly had 315,041 genes, with at most 23,396 possible candidates (7.43% of genes) for reassignment. The default threshold for treeinform is highlighted in blue.

```{r, fused_percentage, echo=FALSE}
orig_assembly = 315041
t=1:10 # original thresholds were 5000*10^(-t)
threshold = 5000*10^(-t)
# numbers taken from slurm output
reassigned = c(23396, 23396, 22792, 13380, 7944, 5560, 5229, 5219, 5092, 0)
newly_created = c(9009, 9009, 9028, 6006, 3663, 2539, 2381, 2376, 2393, 0)
revised_assembly = c(300654, 300654, 301277, 307667, 310760, 312020, 312193, 312198, 312342, 315041)
fused.df <- data.frame(threshold,reassigned, newly_created, revised_assembly)

# additional thresholds to fill out plot: 0.01, 0.025, 0.035, 0.15, 0.25, 0.35, 1.5, 2.5, 3.5
threshold_add = c(0.01, 0.025, 0.035, 0.15, 0.25, 0.35, 1.5, 2.5, 3.5)
reassigned_add = c(6162, 7029, 7429, 9781, 11090, 12123, 18216, 20426, 21808)
newly_created_add = c(2828, 3236, 3424, 4480, 5040, 5481, 7842, 8542, 8895)
revised_add = c(311707, 311248, 311036, 309740, 308991, 308399, 304667, 303157, 302128)
fused.df <- rbind(fused.df, data.frame(threshold=threshold_add, reassigned=reassigned_add, newly_created=newly_created_add, revised_assembly=revised_add))
fused.df.percents <- transmute(fused.df, threshold, reassigned=reassigned/315041)

# coloring for threshold selection & mixture model selection
grp <- c(rep(1, 4), 2, rep(1, 5), 3, rep(1, 8))
fused.df.percents <- cbind(fused.df.percents, grp)

ggplot(fused.df.percents, aes(threshold, reassigned)) + geom_jitter(aes(colour=grp)) + scale_x_log10() + ggtitle("Percentage of reassigned genes vs threshold") + xlab("Threshold") + ylab("Percentage") + theme_classic() + theme(legend.position="none")
```

## Figure 3b: Theoretical duplication times compared to empirical duplication times from before treeinform and after treeinform

In order to validate that treeinform produces more accurate gene trees, we compare the density of duplication times under a birth-death model with parameter $\lambda$=birth rate, $\mu$=death rate, and $t_{or}=\{ t_{or,1}, t_{or,2}, \ldots, t_{or,K} \}$=time of tree origin for gene families $1, \ldots, K$ to the distribution of estimated duplication times for gene trees from before treeinform, and gene trees from after treeinform under 3 different thresholds: 50, 0.05, and 5e-05.

For the distribution of estimated duplication times of gene trees before and after treeinform, we used phylDOG [@boussau2013genome] to estimate duplication times in units of expected number of substitutions, then calibrated branch lengths to units of time based on the Siphonophora species tree that was calibrated to 1 using the chronos function from ape. (TODO: expand on this)

The pdf of duplication times under a birth-death model and given a time of origin is:
$$
f(x_{i,k} | t_{or,k}, \lambda, \mu) = (\lambda - \mu)^2 \frac{e^{-(\lambda-\mu)x_i}}{(\lambda - \mu e^{-(\lambda - \mu)x_i})^2} \frac{\lambda - \mu e^{-(\lambda - \mu)t}}{1 - e^{-(\lambda - \mu)t}} 
$$
where $x_{i, k}$ represents duplication times $i$ from gene tree $G_k$, as derived by Gernhard 2008. To compute this, we first estimate $\lambda$ and $\mu$ using CAFE3 [@han2013estimating] with the same fixed $t_{or}=1,000,000$ and $G=\{ G_1, \ldots, G_K \}$=gene family sizes for gene trees $1, \ldots, K$, then plug the $\lambda$ and $\mu$ estimates from CAFE3 in along with the same calibrated $t_{or,k}$s. A key assumption here is that the fixed $t_{or,k}$s allows for accurate comparisons between the theoretical density of duplication times and the estimated duplication times even if it is technically an incorrect estimate of the age of the gene trees (and species tree).

```{r, fig2b, echo=FALSE}
dt <- c(dt_before, threshold_5, threshold_2, threshold_8, threshold_mix)
l1 = length(dt_before)
l2 = length(threshold_5)
l3 = length(threshold_2)
l4 = length(threshold_8)
l5 = length(threshold_mix)
#grp <- rep(NA, l1+l2+l3+l4+l5)
Threshold <- rep(NA, l1+l2+l3+l4+l5)
Threshold[1:l1] <- "Before"
Threshold[l1+1:l2] <- "0.05"
Threshold[l1+l2+1:l3] <- "50"
Threshold[l1+l2+l3+1:l4] <- "5e-05"
Threshold[l1+l2+l3+l4+1:l5] <- "0.01"
#grp[l1+l2+l3+l4+1:l5] <- "threshold=0.01"
data = data.frame(dt, Threshold)
f = function(x) dup_pdf(x, 1, 0.617025240076, 0.6334249271090)
ggplot(data, aes(dt)) + geom_density(aes(group=Threshold, color=Threshold)) + stat_function(fun=f) + ggtitle("Density of inferred and theoretical duplication times\nunder different treeinform thresholds") + xlab("Duplication Times") + ylab("Density") + theme_classic()
```

## Table 3c: Kolmogorov-Smirnov statistics for before treeinform and after treeinform

Additionally, we can run the Kolmogorov-Smirnov test on the CDFs of the thresholds to compare. The CDF of the theoretical distribution is:
$$
F(x_{i,k} | t=t_{or, k}, \lambda, \mu) = \frac{1 - e^{-(\lambda - \mu x_{i,k})}}{\lambda - \mu e^{-(\lambda - \mu)x_{i,k}}} \frac{\lambda - \mu e^{-(\lambda - \mu) t}}{1 - e^{-(\lambda - \mu) t}}
$$

```{r, ks, echo=FALSE}
theoretical = function(x) dup_cdf(x, 1, 0.617025240076, 0.6334249271090)
ks.stat <- c(ks.test(dt_before, theoretical)$statistic, ks.test(threshold_2, theoretical)$statistic, ks.test(threshold_5, theoretical)$statistic, ks.test(threshold_8, theoretical)$statistic, ks.test(threshold_mix, theoretical)$statistic)
ks.pval <- c(ks.test(dt_before, theoretical)$p.value, ks.test(threshold_2, theoretical)$p.value, ks.test(threshold_5, theoretical)$p.value, ks.test(threshold_8, theoretical)$p.value, ks.test(threshold_mix, theoretical)$p.value)
ks <- data.frame(ks.stat, ks.pval)
colnames(ks) <- c("Statistic", "P-Value")
rownames(ks) <- c("Before", "50", "0.05", "5e-05", "0.01")
kable(ks)
```

The default threshold (0.05) comes closest to the theoretical distribution.

# References