---
title: Revising transcriptome assemblies with phylogenetic information in Agalma1.0
  - Supplementary Information
output:
  pdf_document:
    toc: yes
  html_notebook: default
  html_document: default
bibliography: supplementary.bib
---

```{r, setup, include=FALSE}
library(ape)
library(scales)
library(ggplot2)
library(dplyr)
library(hutan)
library(treeio)
library(parallel)
library(knitr)
library(runjags)
library(entropy)
#library(forecast) if we want boxcox

cores = detectCores()
if (cores < 1) { cores = 1 }
set.seed(1287234)

source("code/functions.R")
```

```{r, parse_trees, include=FALSE, warnings=FALSE}
# Parse trees before everything else
# make speciation calibration times (taken from chronos)
age <- c(0.1718036, 0.1430510, 0.2490215, 0.3227746, 0.5157844, 1.0000000)
clade <- c("Calycophora", "Agalmatidae", "Chodorophora", "Siphonophora", "Hydrozoa", "Cnidaria")
calibration_times = data.frame(age, clade, stringsAsFactors = FALSE)

#5187 gene trees
#s <- 1:5606
#before <- dt_phyldog(s, "data/revisions/phyldog/before/", calibration_times)
#trees_before <- before[[1]]
#calibrated_before <- before[[2]]
#dt_before <- before[[3]]
#save(dt_before, file="data/revisions/dt_before.Rdata")
#save(trees_before, file="data/revisions/trees_before.Rdata")
#save(calibrated_before, file="data/revisions/calibrated_before.Rdata")
load("data/revisions/dt_before.Rdata")
load("data/revisions/trees_before.Rdata")
load("data/revisions/calibrated_before.Rdata")

#l <- 36523:43609
#threshold_5 <- dt_phyldog(l, "data/revisions/phyldog/threshold-5/", calibration_times)[[3]]
#save(threshold_5, file="data/revisions/dt_threshold-5.Rdata")
load("data/revisions/dt_threshold-5.Rdata")

#l <- 51176:56944
#threshold_7 <- dt_phyldog(l, "data/revisions/phyldog/threshold-7/", calibration_times)[[3]]
#save(threshold_7, file="data/revisions/dt_threshold-7.Rdata")
load("data/revisions/dt_threshold-7.Rdata")

#l <- 58509:65608
#threshold_8 <- dt_phyldog(l, "data/revisions/phyldog/threshold-8/", calibration_times)[[3]]
#save(threshold_8, file="data/revisions/dt_threshold-8.Rdata")
load("data/revisions/dt_threshold-8.Rdata")

#l <- 88484:94021
#threshold_mix <- dt_phyldog(l, "data/revisions/phyldog/mix/", calibration_times)[[3]]
#save(threshold_mix, file="data/revisions/dt_threshold-mix.Rdata")
load("data/revisions/dt_threshold-mix.Rdata")
```

## Supplementary methods

The code for the analyses presented here (including an executable version of this document) are available in a git repository at https://github.com/caseywdunn/ms_treeinform. The phylogenetic analyses considered here are based on a 7 taxon siphonophore [@Dunn2009] dataset. This dataset originated as the regression test dataset for the Agalma automated phylogenetics workflow [@Dunn2013] and was selected for its well-resolved species tree as well as the fact that treeinform is implemented in Agalma. The gene trees were built with Agalma1.1.0, and bash scripts for the run can be found at https://github.com/caseywdunn/ms_treeinform/code/revisions. Most analyses are loaded from saved Rdata objects to speed up reproduction of the notebook, with the code to generate the Rdata objects commented out. To reproduce the notebook in its entirety from the raw Agalma output will take around 5 hours.

The phylogenetic analyses in Agalma followed standard approaches with default settings. Speciation and duplication nodes were identified in the gene trees with phyldog [@boussau2013genome]. Bash scripts and associated code for the runs can be found at in the same directory as above. Agalma uses the transcriptome assembler Trinity [@grabherr-2011], version 2.5.2.

## Assessing the extent of transcript assignment errors

We first examined the prevalence of transcript misassignment. For each node in each of the `r toString(length(trees_before))` gene phylogenies, we calculated the length of the corresponding subtree. This is the sum of the length of all branches in the subtree defined by the node. An excess of very short subtrees would be a strong indication of assigning different transcripts of the same gene, which have very similar sequences and therefore short branches connecting them in phylogenetic trees, to different genes. This is the pattern we found (Supplementary Figure 1).

Two issues could create a misleading impression in the histogram of subtree lengths for internal nodes (Supplementary Figure 1). First, it considers all subtrees, including those defined by both speciation and duplication nodes. Misassigning transcripts from the same gene to multiple genes will artificially inflate only the number of duplication nodes, since variation across transcripts within a gene are essentially misassigned to gene duplication events. Examining just the duplication events in the gene trees therefore provides a more direct perspective on the problem we investigate here. Second, subtree lengths are in units of expected numbers of substitution, which depend on both rates of molecular evolution and time. Because the rates of evolution can vary within and between gene phylogenies, variation in rates could confound the interpretation of gene tree sublength. 

We therefore performed a calibrated analysis and focused only on duplication nodes. We first created a time calibrated species tree, with all tips with age 0 and the root node with age 1. We then transformed the branch lengths of the gene trees so that each speciation node in each gene tree had the same age as the corresponding node in the species tree (see source code for this document). A histogram of the calibrated duplication times (Supplementary Figure 2) indicates there is a large excess of recent duplications. This provides additional evidence for the frequent misassignment of transcripts from the same gene to artefactual recent gene duplicates.

```{r, subtree_histo, echo=FALSE, warning=FALSE, fig.cap="Histogram of subtree lengths for internal nodes in each Siphonophora subset gene tree from Agalma1.0 containing tip descendants from the same species. Subtree lengths greater than 1 were filtered out for clarity."}
lengths = unlist(multi_subtree_lengths(trees_before, cores))

lengths_lim = lengths[which(lengths<1)]

ggplot(data=data.frame(x=lengths_lim)) + geom_histogram(aes(x=x,y=..count../sum(..count..)), binwidth=0.01, fill='white', color='black') + theme_classic() + xlab("Subtree length") + ylab("Frequency") #+ geom_vline(xintercept=0.02, linetype='dashed', color='#56B4E9') + geom_text(data=data.frame(x=0.02,y=0.08), map=aes(x=x, y=y), label="0.02", vjust=1.4, angle=90, colour="#56B4E9", size=3)
```

### Comparison to other transcript clustering methods

In order to get a sense of whether the transcript misassignment errors were localized to Trinity or are a more general problem to trasncriptome assembly, we compared Trinity transcript clustering results with another transcript clustering tool, Corset [@Davidson2014] for the 5 species that had to be assembled. For 3 of the samples (SRX288285, SRX288430, SRX288431) we also ran cd-hit [@Fu2012] to remove transcripts with 100% identity in order to address some speed issues in Corset. The distribution of cluster sizes (Supplementary Figure 2) suggests that Corset tends to overcluster compared to Trinity, which would lead to similar misassignment errors.

```{r, corset_compare, echo=FALSE, warnings=FALSE, fig.cap="Cluster size counts for Trinity assembly and Corset clustering algorithm on Trinity contigs. There are 3 Trinity clusters with size greater than 30, while there are 20 Corset clusters with size greater than 30. This suggests that the same misassignment errors are generated by other transcriptome assemblers and clustering algorithms as well."}
#clusters.mod.txt <- (Sys.glob("data/revisions/corset/*-clusters.mod.txt"))
#corset_files <- lapply(clusters.mod.txt, function(x) read.csv(x, sep="\t", col.names = c("Trinity gene", "Trinity isoform", "Corset gene")))
#save(corset_files, file="data/revisions/corset_clusters.Rdata")
load("data/revisions/corset_clusters.Rdata")

trinity_clusters <- lapply(corset_files, function(x) cluster_size_distribution(x$Trinity.gene))
corset_clusters <- lapply(corset_files, function(x) cluster_size_distribution(x$Corset.gene))
catalog_ids=c('SRX288276','SRX288285','SRX288430','SRX288431','SRX288432')
suppressWarnings(size_dist_dfs<-lapply(1:5, function(x)
  {
  nrx<-nrow(trinity_clusters[[x]])
  nry<-nrow(corset_clusters[[x]])
  nrt<-nrx+nry
  df<-rbind(trinity_clusters[[x]],corset_clusters[[x]])
  tmp <- rep("Trinity", nrt)
  tmp[nrx+1:nry]<-"Corset"
  df$'Method' <- tmp
  df$'ID' <- catalog_ids[x]
  return(df)
  }))
suppressWarnings(new_df <- do.call(rbind, size_dist_dfs))

# basic stats about corset & trinity
trinity_over_30<-filter(new_df, Method=='Trinity' & as.numeric(levels(size))[size]>30)
num_trinity_over_30<-sum(trinity_over_30$freq) # returns 3
corset_over_30<-filter(new_df, Method=='Corset' & as.numeric(levels(size))[size]>30)
num_corset_over_30<-sum(corset_over_30$freq) # returns 20

suppressWarnings(ggplot(data=new_df,aes(x=size,y=freq,color=Method,group=Method))+geom_point(alpha=0.7)+scale_y_log10()+xlab("Cluster Size")+ylab("Count")+ggtitle("Cluster size counts for Trinity & Corset by sample") + theme_classic() + geom_line() + facet_wrap(~ID, ncol=2) + scale_x_discrete(breaks=seq(0,70,by=10)))
```

Additionally, we computed the Adjusted Rand Index (todo: citation) to get a sense of the similarity between the Trinity and Corset clusterings. (Supplementary Table 1) The Adjusted Rand Index computes the proportion of pairs that either both belong to the same cluster, or both belong to different clusters, corrected for chance. Though we have no ground truth here, the ARI suggests that Trinity and Corset clusterings are more similar than dissimilar, and in the case of SRX288285, SRX288430, and SRX288431 are very similar.

```{r, ari, echo=FALSE}
#filtered_corset <- mclapply(corset_files, function(x) filter_singletons(x), mc.cores=cores)
#aris <- mclapply(filtered_corset, function(x) ari(x[[2]],x[[1]]), mc.cores=cores)
#save(aris, file="data/revisions/aris.Rdata")
load("data/revisions/aris.Rdata")
ari_df <- data.frame('Sample'=catalog_ids,"Adjusted Rand Index"=unlist(aris))
kable(ari_df, format='pandoc', caption='Adjusted Rand Index between Trinity and Corset clusterings by sample.')
```

```{r, corset_trees, echo=FALSE, warning=FALSE, fig.cap="Histogram of subtree lengths for internal nodes in each Siphonophora subset gene tree from Agalma1.1 with Corset clusterings containing tip descendants from the same species. Subtree lengths greater than 1 were filtered out for clarity."}
#k<-1:5644
#corset_tree_files <- "data/revisions/corset/phyldog/"
#corset_trees <- mclapply(k, function(x) parse_gene_trees(processTree(paste0(corset_tree_files, x, ".ReconciledTree"))), mc.cores = cores)
#corset_trees <- corset_trees[which(!unlist(mclapply(corset_trees, is.null)))]
#save(corset_trees, file="data/revisions/corset_trees.Rdata")
load("data/revisions/corset_trees.Rdata")
corset_lengths <- unlist(multi_subtree_lengths(corset_trees,cores))
corset_lengths_lim = corset_lengths[which(corset_lengths<1)]

# are corset subtree lengths different from trinity...?
trinity_corset_ks <- ks.test(corset_lengths,lengths)$statistic
rejection_level_tc <- 1.36*sqrt((length(corset_lengths)+length(lengths))/(length(corset_lengths)*length(lengths)))

ggplot(data=data.frame(x=corset_lengths_lim)) + geom_histogram(aes(x=x,y=..count../sum(..count..)), binwidth=0.01, fill='white', color='black') + theme_classic() + xlab("Subtree length") + ylab("Frequency") #+ geom_vline(xintercept=0.02, linetype='dashed', color='#56B4E9') + geom_text(data=data.frame(x=0.02,y=0.08), map=aes(x=x, y=y), label="0.02", vjust=1.4, angle=90, colour="#56B4E9", size=3)
```

Supplementary Figure 3 shows the histogram of subtree lengths for internal nodes in each Siphonophora subset gene tree with Corset clusterings. It shares the same characteristics as in Supplementary Figure 1, and the two-sample Kolmogorov-Smirnov test returns a test statistic of $D=$`r toString(signif(trinity_corset_ks,3))` which is less than the rejection threshold at level 0.05 for the samples, `r toString(signif(rejection_level_tc,3))`. This indicates that regardless of which transcript clustering method is chosen, the same transcript misassignment errors persist. Given the intrinsic challenges of assigning assembled transcripts to genes it is likely that the same misassignment errors are generated by other transcriptome assemblers as well.

## Selecting a threshold for transcript reassignment

```{r, runjags, include=FALSE}
#load("data/dt_before.Rdata")

# Data
Y = dt_before
N = length(dt_before)

# Initial values to get the chains started:
alpha <- 1
beta <- 1
lambda <- 0.571990480546
mu <- 0.544504933034
t <- 1
Constant <- 10

Ones <- rep(1, N)

#results <- run.jags(model="code/runjags.txt", n.chains=3, thin=1) # will take ~1hr
#results <- autorun.jags(model="code/runjags_revision.txt", n.chains=3, thin=1) # will take ~3hrs
#results_extend <- extend.jags(results, add.monitor=c("component_chosen")) # will take ~24hrs
#save(results, file="data/revisions/runjags_results2.Rdata")
load("data/revisions/runjags_results2.Rdata")
sum_results = results$summaries

alpha = sum_results[1,4]
beta = sum_results[2,4]
mu = sum_results[3,4]
lambda = sum_results[4,4]
p1 = sum_results[5,4]
p2 = sum_results[6,4]

mix1 = function(x, p1=p1, alpha=alpha, beta=beta) { p1*dgamma(x,shape=alpha,scale=beta)}
mix2 = function(x, p2=p2, lambda=lambda,mu=mu) { p2*dup_pdf(x,1,lambda,mu)}

load("data/revisions/posterior_component1.Rdata")
df_post = suppressWarnings(data.frame(dt=dt_before,pc1=posterior_component1) %>% filter(pc1>0))
intersection = filter(df_post, pc1 >= 0.95) %>% select(dt) %>% max

#intersection = uniroot(function(x) mix1(x,p1,alpha,beta)-mix2(x,p2,lambda,mu), c(0,0.25))$root
```

```{r, back_calibrate, warnings=FALSE, include=FALSE}
# find duplication events with t<=intersection in calibrated trees
calibrated_thresh = mclapply(calibrated_before, function(x)
  {
  if(is.na(x)) { return(NA) }
  else {
    l = length(x@phylo$tip.label)
    filter(x@data, height <= intersection, node > l) %>% select(node) %>% unlist }
  }, mc.cores = cores)

# compute subtree length for equivalent event in uncalibrated trees
subtree_lengths = mcmapply(function(x, y) {
  if(length(x)==0 || is.na(x)) { return(0) }
  else {
    phylo = y@phylo
    data = y@data
    bl = c(phylo$edge.length, 0)
    data[,'bl'] = bl[order(c(phylo$edge[,2], length(phylo$tip.label)+1))]
    des = lapply(x, function(z) descendants(phylo, z))
    len = lapply(des, function(z) {
      filter(data, node %in% z) %>% select(bl) %>% sum
    })
    return(len)
  }
}, calibrated_thresh, trees_before, mc.cores=cores)

# select max out of subtree lengths
thresh = unlist(subtree_lengths) %>% max

stl <- unlist(lengths)
```

A visual inspection of the histogram of subtree lengths (Supplementary Figure 1) suggested that the frequency of subtree length for internal nodes was high below 0.01 but leveled out above it. Numerically, `r toString(signif(length(stl[stl<0.01])/length(stl)*100,4))`% of internal nodes containing tip descendants from the same species had a subtree length less than 0.01, with an additional `r toString(signif(length(stl[stl>0.01 & stl<0.02])/length(stl)*100,4))`% having a subtree length between 0.01 and 0.02. It is unlikely that all of these clades are gene duplication events.

These observations suggest that two different processes are operating simultaneously to generate the observed subtree lengths, one for the misassigned transcripts and one for the correctly assigned transcripts. To model this pattern, we applied a mixture model to the inferred duplication times (equivalent to branch lengths; see main text) from the gene trees. One component modelled duplication events and associated times arising from transcripts assigned to different genes that belong to the same gene (i.e., misassigned transcripts), and the other component modelled duplication events and associated times arising from transcripts assigned to different genes that in fact do belong to different genes (i.e., corrrectly assigned transcripts) (Supplementary Figure 4).

We expected the implied duplication events of transcripts of the same gene that were misassigned to different genes to have very short implied duplication times approaching 0, and thus chose to model that component (Component 1) as a gamma distribution with parameters shape$=\alpha$ and rate$=\beta$. To model duplication events and associated times arising from the correctly assigned transcripts (Component 2), we used a birth-death process [@gernhard2008conditioned], which is well studied and often applied to gene analyses of duplication and loss. The probability distribution function in the model we used has parameters birth rate $\lambda$, death rate $\mu$, and tree time of origin $t_{or}$. Because we fitted a chronogram with time of origin 1 onto the gene trees $G = \{ G_1, G_2, \ldots, G_K \}$, we made the assumption that all gene tree times of origin are $t_{or} = 1$. Some gene trees have duplication events predating the first speciation event, thus when we fitted chronograms onto those gene trees they had times of origin greater than 1. We chose to filter these gene trees out of the mixture model and subsequent analyses.

Let $x_{i, k}$ represent duplication time $i$ from gene tree $G_k$, with $z_{i} \in \{ 1, 2 \}$ representing whether $x_{i,k}$ is drawn from the 1st component $(z_i = 1)$ or the 2nd component $(z_i = 2)$. Then if $\pi_1$ and $\pi_2$ denote the overall probability that a duplication time belongs to the 1st and 2nd component respectively, $\Gamma(x_{i,k} | \alpha, \beta)$ is the probability density function for the gamma distribution, and $f(x_{i,k} | t_{or,k} = 1, \lambda, \mu)$, we get the expression

$$ P(x_{i,k}) = \pi_1 \Gamma(x_{i,k}|\alpha, \beta) + \pi_2 f(x_{i,k} | t_{or,k}=t, \lambda, \mu)$$

We used Just Another Gibbs Sampler (JAGS) [@Plummer03jags] to perform Bayesian Gibbs Sampling in order to infer the parameters $\alpha$, $\beta$, $\lambda$, and $\mu$ as well as the mixing proportions $\pi_1$ and $\pi_2$. This gave us the parameter estimates in Table 2, as well as the posterior probability that a duplication time $x$ is drawn from component 1 or component 2, i.e. $P(z|x)$.

```{r, runjags_table, echo=FALSE}
sum_format = sum_results[,c("Lower95", "Mean", "Upper95", "MCerr")]
rownames(sum_format) <- c("$\\alpha$", "$\\beta$", "$\\mu$", "$\\lambda$", "$\\pi_1$", "$\\pi_2$")
kable(sum_format, format='pandoc', caption='Summary of parameter estimates from JAGS.')
```

The posterior probability gives us a way to determine the probability of error, as $P(z = 1 | x)$ is the error if we decide $x$ is drawn from the 2nd component, and $P(z = 2 | x)$ is the error if we decide $x$ is drawn from the 1st component. If we care more about having fewer correctly assigned transcripts being erroneously flagged as misassigned, then we can use the posterior probability to select an appropriate threshold for treeinform by selecting $T$ such that $P(z = 2 | x) < \alpha$ for all $x < T$, where $\alpha$ is the error rate. Here we decided to use $\alpha = 0.05$. In Bayesian decision theory this is equivalent to a loss matrix of $\begin{bmatrix} \lambda_{11} & \lambda_{12} \\ \lambda_{21} & \lambda_{22} \end{bmatrix} = \begin{bmatrix} 0 & 1 \\ 19 & 0 \end{bmatrix}$, where each entry $\lambda_{mn}$ is the penalty for selecting component $n$ when $x_{i,k}$ is actually drawn from component $m$. This gives us the threshold `r toString(signif(intersection,4))`.

From that threshold, we can back-calibrate to determine a subtree branch length threshold for use in treeinform. Specifically, we took all duplication events with times below the intersection point on all chronogram-fitted gene trees, mapped them to the equivalent events on the phyldog-outputted gene trees, computed the subtree length of all events, and then took the maximum of those subtree lengths. From the intersection point `r toString(signif(intersection,4))`, this gave us a threshold of `r toString(signif(thresh,3))`.

```{r, runjags_viz, echo=FALSE, warnings=FALSE, fig.cap='Histogram of the inferred duplication times. We first ran phyldog [@boussau2013genome] on the Siphonophora subset multiple sequence alignments from Agalma1.1.0 and a user-inputted species tree. This provided gene trees with internal nodes annotated as duplication or speciation events. We then fitted chronograms onto these gene trees with our user-inputted species tree.'}
#Freedman-Diaconis rule for binwidth ~= 0.025 (to make it even)
#bw <- 2*IQR(dt_before)/length(dt_before)^(1/3)

ggplot(data.frame(x=dt_before)) + geom_histogram(aes(x=x,y=..density..), fill='white', color='black', binwidth=0.01) + stat_function(geom="line",fun=mix1,args=list(p1,alpha,beta),xlim=c(0.005,1), aes(color="line1"))+stat_function(geom="line",fun=mix2,args=list(p2,lambda,mu),aes(color="line2"))+ggtitle("Density Curves of Mixture Model Plotted on Histogram of\n Inferred Duplication Times Before Treeinform") + scale_color_manual("Components of Mixture Model", labels=c('line1' = expression("Component 1, " * pi[1]*Gamma(x, alpha, beta)), 'line2' = expression("Component 2, "*pi[2]*f(x,1,lambda,mu))), values=c("#FF000080", "blue")) + xlab("Duplication Times") + ylab("Density") + theme_classic() + theme(legend.position = c(0.8, 0.6))

#z = theme(legend.position = "none", text=element_text(colour='#838787', size = 20), plot.background=element_rect(fill ='#222222'), panel.background=element_rect(fill='#222222'), legend.background=element_rect(fill='#222222'), axis.line=element_line(colour='#838787'), axis.text=element_text(colour='#838787', size=15), axis.ticks = element_line(colour='#838787'), line=element_line(size=5))
#ggplot(data.frame(x=dt_before)) + geom_histogram(aes(x=x,y=..density..), fill='#838787', color='black', binwidth=0.01) + geom_vline(xintercept=0.01, linetype='dashed', color='yellow') + geom_text(data=data.frame(x=0.01,y=5), map=aes(x=x, y=y), label="0.01", vjust=1.4, angle=90, colour="yellow", size=8) + stat_function(geom="line",fun=mix1,args=list(p1,alpha,beta),xlim=c(0.003,1), colour="#E42832", size=1)+stat_function(geom="line",fun=mix2,args=list(p2,lambda,mu),colour="#34A5DA", size=1) + ylab("Density") + xlab("Time") + theme_classic() + z
```

## Validating the effectiveness of treeinform

```{r, fused_percentage, echo=FALSE, fig.cap='The percentage of reassigned tips is plotted above on a log scale. The original assembly had 433,071 genes, with at most 23,396 possible candidates (7.43% of genes) for reassignment. The default threshold for treeinform is highlighted in light blue (TODO: its not, yet).'}
orig_assembly = 433071
t=1:10 # original thresholds were 5000*10^(-t)
threshold = 5000*10^(-t)
# numbers taken from slurm output
reassigned = c(23560, 23560, 22930, 11515, 5350, 3364, 3052, 3024, 2955, 0)
newly_created = c(9167, 9167, 9189, 5271, 2495, 1548, 1398, 1384, 1395, 0)
revised_assembly = c(418678, 418678, 419330, 426827, 430216, 431255, 431417, 431431, 431511, 433071)
fused.df <- data.frame(threshold,reassigned, newly_created, revised_assembly)

# additional thresholds to fill out plot: 0.01, 0.025, 0.035, 0.15, 0.25, 0.35, 1.5, 2.5, 3.5
threshold_add = c(0.01, 0.02, 0.035, 0.15, 0.25, 0.35)
reassigned_add = c(3762, 4333, 4861, 7371, 8877, 10098)
newly_created_add = c(1736, 2008, 2262, 3430, 4108, 4647)
revised_add = c(431045, 430746, 430472, 429130, 428302, 427620)
fused.df <- rbind(fused.df, data.frame(threshold=threshold_add, reassigned=reassigned_add, newly_created=newly_created_add, revised_assembly=revised_add))
fused.df.percents <- transmute(fused.df, threshold, reassigned=reassigned/orig_assembly)

# coloring for threshold selection & mixture model selection
#grp <- c(rep(1, 18), 2)
#fused.df.percents <- cbind(fused.df.percents, grp)

#ggplot(fused.df.percents, aes(threshold, reassigned)) + geom_jitter(aes(colour=grp)) + scale_x_log10() + ggtitle("Percentage of reassigned genes vs threshold") + xlab("Threshold") + ylab("Percentage") + theme_classic() + theme(legend.position="none") + geom_vline(xintercept=0.02, linetype='dashed', color='#56B4E9') + geom_text(data=data.frame(x=0.02,y=0.015), map=aes(x=x, y=y), label="0.02", hjust=-0.4, colour="#56B4E9", size=3)
ggplot(fused.df.percents, aes(threshold, reassigned)) + geom_jitter() + scale_x_log10() + ggtitle("Percentage of reassigned genes vs threshold") + xlab("Threshold") + ylab("Percentage") + theme_classic() + theme(legend.position="none")
```

In order to validate that treeinform improves the accuracy of assigning transcripts to genes under the specified threshold, we performed two analyses. First, we plotted the percentage of reassigned genes at different thresholds to assess the performance of the default threshold value of `r toString(signif(thresh,3))` (Supplementary Figure 5). Below the default value, the percentage of reassigned genes begins to plateau, while above the default value the percentage of reassigned genes increases very quickly, increasing the likelihood of treeinform to reassign transcripts from different genes to the same gene in addition to reassign transcripts from the same gene together.

Second, we compared the density of duplication times under the model provided for Component 2 of the mixture model to the distribution of estimated duplication times for gene trees from Agalma1.0 before treeinform, and gene trees from Agalma1.0 after treeinform under 3 different thresholds: 0.05, 0.0005, and 0.00005 (Supplementary Figure 6). We again fitted chronograms with the same Siphonophora species tree onto all gene trees from Agalma1.1 and filtered out those gene trees with time of origin greater than 1, so that duplication times were comparable between trees. Visually, the analyses with the 0.02 threshold comes closest to the theoretical. (TODO: edit this section)

```{r, fig3b, echo=FALSE, warnings=FALSE, fig.cap="Density from theoretical and the empirical density under the 3 different thresholds as well as before treeinform was run. The distribution before treeinform has a large peak on the left that is removed by treeinform with all examined thresholds."}
dt <- c(dt_before, threshold_mix, threshold_5, threshold_7, threshold_8)
l1 = length(dt_before)
l2 = length(threshold_mix)
l3 = length(threshold_5)
l4 = length(threshold_7)
l5 = length(threshold_8)
Threshold <- rep(NA, l1+l2+l3+l4+l5)
Threshold[1:l1] <- "Before"
Threshold[l1+1:l2] <- "0.07"
Threshold[l1+l2+1:l3] <- "0.05"
Threshold[l1+l2+l3+1:l4] <- "0.0005"
Threshold[l1+l2+l3+l4+1:l5] <- "0.00005"
data = data.frame(dt, Threshold)
f = function(x) dup_pdf(x, 1, lambda, mu)
ggplot(data, aes(dt)) + geom_density(aes(group=Threshold, color=Threshold), bw=0.02) + stat_function(fun=f) + ggtitle("Density of inferred and theoretical duplication times\nunder different treeinform thresholds") + xlab("Duplication Times") + ylab("Density") + theme_classic() + theme(legend.position = c(0.8, 0.6)) + geom_hline(yintercept=0, color="white") + geom_vline(xintercept=0, color="white")

# FOR PRES
#ggplot(data, aes(dt)) + geom_density(aes(group=Threshold, color=Threshold), bw=0.02) + stat_function(fun=f, color='white') + ggtitle("Density of inferred and theoretical duplication times\nunder different treeinform thresholds") + xlab("Duplication Times") + ylab("Density") + theme_classic() + geom_hline(yintercept=0, color="#222222") + geom_vline(xintercept=0, color="#222222") + z
```

Additionally, we computed the Kullback-Leibler distance [@kullback1951] between the distributions of duplication times under different thresholds and the theoretical distribution of duplication times (Table 2). Kullback-Leibler distance, otherwise known as relative entropy, measures the distance between two distributions. (TODO: update this. Maybe a plot of KL distance would be better?) The KL distance between the distribution of duplication times after running treeinform with the default threshold value of 0.02 come closest to the theoretical distribution as compared to both threshold levels below and above the default value. This indicates that treeinform produces more accurate gene trees with appropriate threshold selection.

```{r, kl, echo=FALSE, warning=FALSE, fig.cap=''}
z.pre <- rep(NA, 10000)
u <- runif(10000)
theoretical = function(x, u) dup_cdf_sampler(x, 1, lambda, mu, u)
my.uniroot <- function(x) uniroot(theoretical, c(0, 1), u=x)$root
r <- vapply(u, my.uniroot, numeric(1))

d_theo <- discretize(r, numBins=100)
d_before <- discretize(dt_before, numBins=100)
d_mix <- discretize(threshold_mix, numBins=100)
#d_2 <- discretize(threshold_2, numBins=100)
d_5 <- discretize(threshold_5, numBins=100)
d_7 <- discretize(threshold_7, numBins=100)
d_8 <- discretize(threshold_8, numBins=100)
kl_dist <- c(KL.empirical(d_before, d_theo), KL.empirical(d_mix, d_theo), KL.empirical(d_5, d_theo), KL.empirical(d_7,d_theo), KL.empirical(d_8, d_theo))

kl <- data.frame("KL Distance" = kl_dist)
rownames(kl) <- c("Before", "0.07", "0.05", "0.0005", "5e-05")
kable(kl, format='pandoc', caption='Kullback-Leibler distances between duplication times after running treeinform with different thresholds and theoretical duplication times.')
```

## Software versions

This manuscript was computed on `r format( Sys.time(), "%a %b %d %X %Y" )` with the following R package versions.

```{r session_summary, echo=FALSE, comment=NA}
	sessionInfo()
```

\pagebreak


## References
